DIRKRB-692 - populates SGT clientPrincipal and fixes storeTicket method

ftarlao opened the pull request:
For details refer to the issue:
https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

ftarlao committed:
Fix for SGT clientPrincipal that is currently not populated. This fix
should be moved at a lower layer but this is a quick fix that works (tested with USE_TGT case)
REF DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

ftarlao committed:
Fix for storeTicket method, it does not support correctly the one SGT only
case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java:
```diff
@@ -271,15 +271,23 @@ public void storeTicket(TgtTicket tgtTicket,
-                cCache.addCredential(new Credential(sgtTicket, sgtTicket.getClientPrincipal()));
-                cCache.setPrimaryPrincipal(sgtTicket.getClientPrincipal());
+                CredentialCache cCache;
+                if(!isFreshNew){
```
Please put spaces before "(" and after ")".

ftarlao replied:
I'll fix

ftarlao replied:
I should also improve the condition: isFreshNew = !ccacheFile.exists() || (ccacheFile.length() == 0); in order to manage corner cases (opening a file have caused me problems in the past about automatic file creation... of empty files, don't know if this is OS related).

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java:
```diff
@@ -271,15 +271,23 @@ public void storeTicket(TgtTicket tgtTicket,
+                CredentialCache cCache;
+                if(!isFreshNew){
+                    //Remind, contructor sets the cCache client principal from the sgtTicket one
+                    cCache = new CredentialCache(sgtTicket); 
```
Why not "new CredentialCache()" here as the code was before?

ftarlao replied:
right..

ftarlao replied:
I should also remove, in the same block, the cCache.setPrimaryPrincipal(sgtTicket.getClientPrincipal()); because adding an SGT should not change the primary client principal.. I think

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java:
```diff
@@ -153,13 +153,14 @@ public SgtTicket requestSgt(KOptions requestOptions) throws KrbException {
                 KrbConfig krbConfig = krbSetting.getKrbConfig();
                 LinkedList<String> capath = krbConfig.getCapath(sourceRealm, destRealm);
-                PrincipalName clientPrincipalName = tgtTicket.getClientPrincipal();
+                clientPrincipalName = tgtTicket.getClientPrincipal();
```
This line doesn't apply if the source realm is equal to the dest realm, so you could move it above the if statement.

ftarlao replied:
right, I'll fix

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java:
```diff
@@ -170,11 +171,19 @@ public SgtTicket requestSgt(KOptions requestOptions) throws KrbException {
                 }
             }
+        } else {
+            //This code is for the no-tgt case but should be double-checked
```
Is this part required for your use-case? If not I think it could be left out.

ftarlao replied:
I think that trying to support a specific 'manage use-case' is wrong (IMHO), I always try to manage all the use-cases :-) . I have a question, is there a way to request an sgt without tgt? (e.g., OTP, user/pass) and in that case, where is the principal name stored?  I have speculated (in the KOptions) CLIENT_PRINCIPAL. Or is the USE_TGT the only accepted case for requestSGT?

ftarlao replied:
At least, I'll leave an empty 'else' block with a reminder .. for other possible use cases...

ftarlao replied:
I also improved in another comments the clientPrincipal "null" case.. please refer to commit comments for details

ftarlao committed:
Improves previous fix for requestSGT method, small bug fix, typos and 
improved comment
case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

ftarlao committed:
Improves previous fix for storeTicket method, fixed behaviour of 
no-fresh-new case and improved comments +  better formatting
case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

ftarlao committed:
Improves previous fix for requestsgt method, null clientPrincipal is not
saved in sgt, this will preseve values coming from lower layers

case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java:
```diff
@@ -271,15 +271,22 @@ public void storeTicket(TgtTicket tgtTicket,
             try {
-                cCache.load(ccacheFile);
-                cCache.addCredential(new Credential(sgtTicket, sgtTicket.getClientPrincipal()));
-                cCache.setPrimaryPrincipal(sgtTicket.getClientPrincipal());
```
Why is this line removed?

ftarlao replied:
When ccache exists and is correctly loaded, the cCache primaryPrincipal field should be already populated. Furthermore, I think that adding an SGT should not change the cCache primary principal (when present). I think that, setting the primary principal, is only needed when  fresh new cache file is created.

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/impl/AbstractInternalKrbClient.java:
```diff
@@ -170,11 +171,22 @@ public SgtTicket requestSgt(KOptions requestOptions) throws KrbException {
         tgsRequest.setRequestOptions(requestOptions);
-        return doRequestSgt(tgsRequest);
+        SgtTicket sgtTicket = doRequestSgt(tgsRequest);
+        if(clientPrincipalName!=null){
```
Please put spaces around brackets here.

ftarlao replied:
Gosh, I hate formatting :-) .. I'll fix this

ftarlao committed:
requestsgt method, better formatting 

case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java:
```diff
@@ -271,15 +271,22 @@ public void storeTicket(TgtTicket tgtTicket,
     public void storeTicket(SgtTicket sgtTicket, File ccacheFile) throws KrbException {
         LOG.info("Storing the sgt to the credential cache file.");
-        if (!ccacheFile.exists()) {
+        boolean isFreshNew = !ccacheFile.exists() || (ccacheFile.length() == 0);
```
Please change "isFreshNew" to "createCache".

ftarlao replied:
done

coheigea commented on kerby-kerb/kerb-client/src/main/java/org/apache/kerby/kerberos/kerb/client/KrbClientBase.java:
```diff
@@ -271,15 +271,22 @@ public void storeTicket(TgtTicket tgtTicket,
         }
         if (ccacheFile.exists() && ccacheFile.canWrite()) {
-            CredentialCache cCache = new CredentialCache();
+            
```
There is some trailing whitespace here, and in a few other places as well that could be removed.

ftarlao replied:
I'm going to commit the changes, removed the blank rows.. I left only the blank rows that were in the original code.

ftarlao replied:
Ok, I totally misunderstood. Trailing spaces are not empty lines! :-( English is not my natural language and sometimes I fail. I added a new commit to the pull request, deleted the trailing spaces and re-added few blank lines around if/else blocks.. that.. I think ... improve readability.
Regards

ftarlao committed:
Deleted trailing spaces and variable "isFreshNew" name refactoring
case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

ftarlao committed:
Deleted trailing spaces (for real) added blank lines around few if/else blocks
case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

coheigea commented:
There is still a lot of trailing whitespace that I can see when I apply the patch. For example, line 279 of KrbClientBase.java.

ftarlao committed:
Deleted (missed) trailing spaces. 
case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

ftarlao committed:
Deleted (other missed) trailing spaces. 
case. REF: DIRKRB-692 https://issues.apache.org/jira/projects/DIRKRB/issues/DIRKRB-692?filter=allopenissues

ftarlao commented:
I'm sorry, I'm not so good about formatting.. I think you have some
automated check for this and that you havee strict rules.
I have tried now with a search regex.. hope to have solved all
trailing-space-issues

On 12 February 2018 at 11:44, Colm O hEigeartaigh <notifications@github.com>
wrote:

> There is still a lot of trailing whitespace that I can see when I apply
> the patch. For example, line 279 of KrbClientBase.java.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/directory-kerby/pull/29#issuecomment-364886476>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD8v3OilFlpvmPkDFHHLOJRzdEV6Lox-ks5tUBYigaJpZM4R8ebQ>
> .
>
